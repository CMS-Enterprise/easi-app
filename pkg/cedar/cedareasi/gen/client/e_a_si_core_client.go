// Code generated by go-swagger; DO NOT EDIT.

package client

// This file was generated by the swagger tool.
// Editing this file might prove futile when you re-run the swagger generate command

import (
	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"

	strfmt "github.com/go-openapi/strfmt"

	"github.com/cmsgov/easi-app/pkg/cedar/cedareasi/gen/client/operations"
)

// Default e a si core HTTP client.
var Default = NewHTTPClient(nil)

const (
	// DefaultHost is the default Host
	// found in Meta (info) section of spec file
	DefaultHost string = "cdrdevapigw1.awscloud.cms.local:5555"
	// DefaultBasePath is the default BasePath
	// found in Meta (info) section of spec file
	DefaultBasePath string = "/gateway/EASi Core API/1.0"
)

// DefaultSchemes are the default schemes found in Meta (info) section of spec file
var DefaultSchemes = []string{"http"}

// NewHTTPClient creates a new e a si core HTTP client.
func NewHTTPClient(formats strfmt.Registry) *EASiCore {
	return NewHTTPClientWithConfig(formats, nil)
}

// NewHTTPClientWithConfig creates a new e a si core HTTP client,
// using a customizable transport config.
func NewHTTPClientWithConfig(formats strfmt.Registry, cfg *TransportConfig) *EASiCore {
	// ensure nullable parameters have default
	if cfg == nil {
		cfg = DefaultTransportConfig()
	}

	// create transport and client
	transport := httptransport.New(cfg.Host, cfg.BasePath, cfg.Schemes)
	return New(transport, formats)
}

// New creates a new e a si core client
func New(transport runtime.ClientTransport, formats strfmt.Registry) *EASiCore {
	// ensure nullable parameters have default
	if formats == nil {
		formats = strfmt.Default
	}

	cli := new(EASiCore)
	cli.Transport = transport

	cli.Operations = operations.New(transport, formats)

	return cli
}

// DefaultTransportConfig creates a TransportConfig with the
// default settings taken from the meta section of the spec file.
func DefaultTransportConfig() *TransportConfig {
	return &TransportConfig{
		Host:     DefaultHost,
		BasePath: DefaultBasePath,
		Schemes:  DefaultSchemes,
	}
}

// TransportConfig contains the transport related info,
// found in the meta section of the spec file.
type TransportConfig struct {
	Host     string
	BasePath string
	Schemes  []string
}

// WithHost overrides the default host,
// provided by the meta section of the spec file.
func (cfg *TransportConfig) WithHost(host string) *TransportConfig {
	cfg.Host = host
	return cfg
}

// WithBasePath overrides the default basePath,
// provided by the meta section of the spec file.
func (cfg *TransportConfig) WithBasePath(basePath string) *TransportConfig {
	cfg.BasePath = basePath
	return cfg
}

// WithSchemes overrides the default schemes,
// provided by the meta section of the spec file.
func (cfg *TransportConfig) WithSchemes(schemes []string) *TransportConfig {
	cfg.Schemes = schemes
	return cfg
}

// EASiCore is a client for e a si core
type EASiCore struct {
	Operations *operations.Client

	Transport runtime.ClientTransport
}

// SetTransport changes the transport on the client and all its subresources
func (c *EASiCore) SetTransport(transport runtime.ClientTransport) {
	c.Transport = transport

	c.Operations.SetTransport(transport)

}
