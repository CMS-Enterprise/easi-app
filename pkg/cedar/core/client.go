package cedarcore

import (
	"context"
	"net/http"
	"time"

	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"
	"github.com/go-openapi/strfmt"
	cache "github.com/patrickmn/go-cache"
	"go.uber.org/zap"

	"github.com/cmsgov/easi-app/pkg/appcontext"
	apiclient "github.com/cmsgov/easi-app/pkg/cedar/core/gen/client"
)

const (
	cedarCoreCacheDurationDefault = time.Hour * 6
)

type loggingTransport struct {
	logger *zap.Logger
}

func (t *loggingTransport) RoundTrip(r *http.Request) (*http.Response, error) {
	start := time.Now().UnixMilli()
	resp, err := http.DefaultTransport.RoundTrip(r)
	end := time.Now().UnixMilli()

	// Start a status code of 0, in case the request fails (and we get a nil resp)
	status := 0
	if resp != nil {
		status = resp.StatusCode
	}

	t.logger.Info(
		"Call to CEDAR core",
		zap.String("service", "cedarcore"),
		zap.Bool("cacheEnabled", false),
		zap.String("method", r.Method),
		zap.Int("status", status),
		zap.String("path", r.URL.Path),
		zap.String("queryParams", r.URL.RawQuery),
		zap.Int64("timeMS", end-start),
	)

	return resp, err
}

// NewClient builds the type that holds a connection to the CEDAR Core API
func NewClient(ctx context.Context, cedarHost string, cedarAPIKey string, cedarAPIVersion string, cacheRefreshTime time.Duration, mockEnabled bool) *Client {
	c := cache.New(cache.NoExpiration, cache.NoExpiration) // Don't expire data _or_ clean it up

	hc := http.Client{
		Transport: &loggingTransport{
			logger: appcontext.ZLogger(ctx),
		},
	}

	basePath := "/gateway/CEDAR Core API/" + cedarAPIVersion
	client := &Client{
		mockEnabled: mockEnabled,
		auth: httptransport.APIKeyAuth(
			"x-Gateway-APIKey",
			"header",
			cedarAPIKey,
		),
		sdk: apiclient.New(
			httptransport.New(
				cedarHost,
				basePath,
				apiclient.DefaultSchemes,
			),
			strfmt.Default,
		),
		hc:    &hc,
		cache: c,
	}

	if cacheRefreshTime <= 0 {
		appcontext.ZLogger(ctx).Error(
			"CEDAR cacheRefreshTime is not a valid value, falling back to default",
			zap.Duration("cacheRefreshTime", cacheRefreshTime),
			zap.Duration("defaultDuration", cedarCoreCacheDurationDefault),
		)
		cacheRefreshTime = cedarCoreCacheDurationDefault
	}
	client.startCacheRefresh(ctx, cacheRefreshTime, client.populateSystemSummaryCache)

	return client
}

// Client represents a connection to the CEDAR Core API
type Client struct {
	mockEnabled bool
	auth        runtime.ClientAuthInfoWriter
	sdk         *apiclient.CEDARCoreAPI
	hc          *http.Client
	cache       *cache.Cache
}

// startCacheRefresh starts a goroutine that will run `populateCache` based on cacheRefreshTime.
// startCacheRefresh returns no errors, and only logs when something goes wrong.
// Upon being called, startCacheRefresh will populate the cache once immediately, then again at an interval specificed by cacheRefreshTime.
func (c *Client) startCacheRefresh(ctx context.Context, cacheRefreshTime time.Duration, populateCache func(context.Context) error) {
	ticker := time.NewTicker(cacheRefreshTime)
	go func(ctx context.Context) {
		for {
			err := populateCache(ctx)
			if err != nil {
				// This is a Warn() instead of Error() as it happens somewhat frequently, and doesn't necessarily warrant immediate attention
				appcontext.ZLogger(ctx).Warn("Failed to refresh cache", zap.Error(err))
			}
			// Wait for the ticker. This will block the current goroutine until the ticker sends a message over the channel
			<-ticker.C
		}
	}(ctx)
}
