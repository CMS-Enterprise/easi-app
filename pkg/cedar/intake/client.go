// Package intake serves as an Anti-Corruption Layer (in Domain Driven Design
// parlance), keeping the implementation details of working with the
// downstream CEDAR Intake API well separated, so the rest of the internal
// workings of the EASi code are not using anything defined by the
// external API
package intake

import (
	"context"
	"fmt"
	"net/http"
	"strconv"
	"time"

	"github.com/go-openapi/runtime"
	httptransport "github.com/go-openapi/runtime/client"
	"github.com/go-openapi/strfmt"
	"go.uber.org/zap"

	"github.com/cms-enterprise/easi-app/pkg/appcontext"
	apiclient "github.com/cms-enterprise/easi-app/pkg/cedar/intake/gen/client"
	"github.com/cms-enterprise/easi-app/pkg/cedar/intake/gen/client/health_check"
	"github.com/cms-enterprise/easi-app/pkg/cedar/intake/gen/client/intake"
	"github.com/cms-enterprise/easi-app/pkg/cedar/intake/translation"
	"github.com/cms-enterprise/easi-app/pkg/models"
	"github.com/cms-enterprise/easi-app/pkg/storage"
)

// NewClient builds the type that holds a connection to the CEDAR Intake API
func NewClient(cedarHost string, cedarAPIKey string, enabled bool, publisherEnabled bool) *Client {
	hc := http.DefaultClient

	return &Client{
		enabled:          enabled,
		publisherEnabled: publisherEnabled,
		auth: httptransport.APIKeyAuth(
			"x-Gateway-APIKey",
			"header",
			cedarAPIKey,
		),
		sdk: apiclient.New(
			httptransport.New(
				cedarHost,
				apiclient.DefaultBasePath,
				apiclient.DefaultSchemes,
			),
			strfmt.Default,
		),
		hc: hc,
	}
}

// Client represents a connection to the CEDAR Intake API
type Client struct {
	enabled          bool
	publisherEnabled bool
	auth             runtime.ClientAuthInfoWriter
	sdk              *apiclient.CEDARIntake
	hc               *http.Client
}

// CheckConnection hits the CEDAR Intake API `/healthcheck` endpoint to verify
// that our connection is functional
func (c *Client) CheckConnection(ctx context.Context) error {
	// If Intake API calls are not enabled, we don't need to check the connection, so just return nil
	if !c.enabled {
		return nil
	}

	params := health_check.NewHealthCheckParamsWithContext(ctx)
	params.HTTPClient = c.hc

	resp, err := c.sdk.HealthCheck.HealthCheck(params, c.auth)
	if err != nil {
		return err
	}
	if resp.Payload == nil {
		return fmt.Errorf("no response body received")
	}
	return nil
}

// PublishSystemIntake sends a system intake to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishSystemIntake(ctx context.Context, si models.SystemIntake) error {
	intakeObject := translation.TranslatableSystemIntake(si)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishBusinessCase sends a business case to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishBusinessCase(ctx context.Context, bc models.BusinessCaseWithCosts) error {
	intakeObject := translation.TranslatableBusinessCase(bc)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishGRTFeedback sends an item of GRT feedback to CEDAR through the Intake API for eventual storage in Alfabet
// func (c *Client) PublishGRTFeedback(ctx context.Context, feedback models.GRTFeedback) error {
// 	intakeObject := translation.TranslatableFeedback(feedback)
// 	return c.publishIntakeObject(ctx, &intakeObject)
// }

// PublishAction sends an action to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishAction(ctx context.Context, action models.Action) error {
	intakeObject := translation.TranslatableAction(action)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// PublishNote sends a note to CEDAR through the Intake API for eventual storage in Alfabet
func (c *Client) PublishNote(ctx context.Context, note models.SystemIntakeNote) error {
	intakeObject := translation.TranslatableNote(note)
	return c.publishIntakeObject(ctx, &intakeObject)
}

// private method for publishing anything that satisfies the translation.IntakeObject interface to CEDAR through the Intake API
func (c *Client) publishIntakeObject(ctx context.Context, model translation.IntakeObject) error {
	// constant values for now; may become non-constant if/when we revisit handling CEDAR validation errors

	// from the Swagger (cedar_intake.json, definitions/IntakeInput/properties/version):
	// "The version associated with the object in the body. This value can be incremented in the event a transaction needs to be resubmitted."
	const objectVersion = 1

	// from the Swagger (cedar_intake.json, paths/"/intake"/post/parameters/validatePayload):
	// "Determines if schema validation of the payload is performed syncronously before persisting the record or asyncronously after the record has been persisted"
	const isValidatedSynchronously = false // false for V1 schemas; this may change when we implement V2 schemas for EASI-1614

	id := model.ObjectID()
	objectType := model.ObjectType()

	if !c.enabled {
		appcontext.ZLogger(ctx).Info("snapshot publishing is not enabled", zap.String("object ID", id), zap.String("object type", objectType))
		return nil
	}

	input, err := model.CreateIntakeModel(ctx)

	if err != nil {
		appcontext.ZLogger(ctx).Error(
			"problem building cedar payload",
			zap.String("object ID", id),
			zap.String("object type", objectType),
			zap.Error(err),
		)
		return nil
	}

	params := intake.NewIntakeAddParamsWithContext(ctx)
	params.HTTPClient = c.hc
	params.Body = input

	params.ValidatePayload = strconv.FormatBool(isValidatedSynchronously)

	objectVersionStr := strconv.FormatInt(objectVersion, 10)
	params.Body.Version = &objectVersionStr

	resp, err := c.sdk.Intake.IntakeAdd(params, c.auth)
	if err != nil {
		return err
	}

	// TODO: we may need to read through the results for any "duplicate entry" error codes
	// that CEDAR _may_ send us, especially if we are re-transmitting entities... We
	// don't really care if there are dupes on their side, because we do NOT want to hold
	// any state on our side about who/what/when may have transmitted a previous copy
	// of the info to CEDAR.
	// appcontext.ZLogger(ctx).Info(fmt.Sprintf("results: %v", resp.Payload))
	_ = resp

	return nil
}

func (c *Client) PublishOnSchedule(ctx context.Context, store *storage.Store, dayOfWeek time.Weekday, hourInUTC int) {
	logger := appcontext.ZLogger(ctx)
	if hourInUTC > 24 || hourInUTC < 0 {
		logger.Error("incorrect hour given for publish schedule, use int between 0 and 24")
		return
	}
	// check both if cedar or the publisher is enabled. If not, return.
	if !c.enabled || !c.publisherEnabled {
		logger.Info("CEDAR intake publish schedule is disabled")
		return
	}

	for {
		nextPublish := getDurationUntilNextDayAndTime(time.Now().UTC(), dayOfWeek, hourInUTC)
		logger.Info(fmt.Sprintf(
			"next intake publish to CEDAR in %s at %s",
			nextPublish,
			time.Now().UTC().Add(nextPublish).Format(time.RFC3339),
		))
		time.Sleep(nextPublish)

		logger.Info("running scheduled intake publish to CEDAR")
		c.publishIntakeAndBusinessCase(ctx, store)
	}
}

func (c *Client) publishIntakeAndBusinessCase(ctx context.Context, store *storage.Store) {
	allIntakes, err := store.FetchSystemIntakes(ctx)
	if err != nil {
		appcontext.ZLogger(ctx).Warn("failed to fetch intakes when publishing to CEDAR", zap.Error(err))
		return
	}
	for _, intake := range allIntakes {
		err = c.PublishSystemIntake(ctx, intake)
		if err != nil {
			appcontext.ZLogger(ctx).Warn(
				"failed to publish intake when publishing to CEDAR",
				zap.String("id", intake.ID.String()),
				zap.Error(err),
			)
		}
		if intake.BusinessCaseID == nil {
			continue
		}
		businessCase, err := store.FetchBusinessCaseByID(ctx, *intake.BusinessCaseID)
		appcontext.ZLogger(ctx).Warn(
			"failed to fetch business case for intake when publishing to CEDAR",
			zap.String("id", intake.ID.String()),
			zap.Error(err),
		)
		if businessCase == nil {
			continue
		}
		err = c.PublishBusinessCase(ctx, *businessCase)
		if err != nil {
			appcontext.ZLogger(ctx).Warn(
				"failed to publish business case to CEDAR",
				zap.String("id", intake.ID.String()),
				zap.Error(err),
			)
		}
	}
}

// This function returns the duration from the start time to a given day of week and hour
func getDurationUntilNextDayAndTime(startTime time.Time, d time.Weekday, hour int) time.Duration {
	daysUntilTarget := getDaysTil(startTime.Weekday(), d)

	// given weekday matches start date, check if the time has passed given hour
	// if it has, add 7 days of duration
	if daysUntilTarget == 0 && startTime.After(getTimeAtHour(startTime, hour)) {
		daysUntilTarget = 7
	}

	// adds days to start time
	targetDay := startTime.AddDate(0, 0, daysUntilTarget)
	// sets hour of day on target day to given hour
	targetDayAndTime := getTimeAtHour(targetDay, hour)
	// return the duration (difference) between the start date and the given day and time
	return targetDayAndTime.Sub(startTime)
}

// This function gets the number of days until the given weekday from the start date
// getDaysTil(dateThatsAThursday.Weekday(), time.Friday) -> returns 1
// getDaysTil(time.Friday, time.Friday) -> returns 0
// getDaysTil(time.Saturday, time.Friday) -> returns 6
func getDaysTil(startDay time.Weekday, targetDay time.Weekday) int {
	return (int(targetDay) - int(startDay) + 7) % 7
}

// This function first sets the time to midnight through Truncate
// and then adds the provided hour to that time
// getTimeAtHour(date, 13) -> returns date with time at 1pm
func getTimeAtHour(t time.Time, hour int) time.Time {
	return t.Truncate(24 * time.Hour).Add(time.Duration(hour) * time.Hour)
}
