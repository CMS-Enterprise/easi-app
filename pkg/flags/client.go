package flags

import (
	"time"

	ld "gopkg.in/launchdarkly/go-server-sdk.v4"

	"github.com/cmsgov/easi-app/pkg/appconfig"
)

// Config has all the parts for creating a new LD Client
type Config struct {
	Source  appconfig.FlagSourceOption
	Key     string
	Timeout time.Duration
	Offline bool
}

// FlagValues is the struct for our keys and values from LD
type FlagValues map[string]interface{}

// NewLaunchDarklyClient returns a client backed by Launch Darkly
func NewLaunchDarklyClient(config Config) (*ld.LDClient, error) {
	ldConfig := ld.Config{Offline: config.Offline}
	return ld.MakeCustomClient(config.Key, ldConfig, config.Timeout)
}
