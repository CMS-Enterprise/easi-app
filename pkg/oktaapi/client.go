// package oktaapi

// import (
// 	"context"

// 	"github.com/cmsgov/easi-app/pkg/models"
// )

// // Client defines methods which any Okta API Client should implement
// type Client interface {
// 	FetchUserInfo(context.Context, string) (*models.UserInfo, error)
// 	SearchByName(context.Context, string) ([]*models.UserInfo, error)
// }
