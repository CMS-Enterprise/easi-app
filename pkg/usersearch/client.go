package usersearch

import (
	"context"

	"github.com/cmsgov/easi-app/pkg/models"
)

// Client is a generic interface for userSearch API
type Client interface {
	FetchUserInfo(context.Context, string) (*models.UserInfo, error)
	FetchUserInfos(context.Context, []string) ([]*models.UserInfo, error)
	SearchCommonNameContains(context.Context, string) ([]*models.UserInfo, error)
}
