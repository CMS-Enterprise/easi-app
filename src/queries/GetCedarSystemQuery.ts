import { gql } from '@apollo/client';

export default gql`
  query GetCedarSystem($id: String!) {
    cedarSystem(cedarSystemId: $id) {
      id
      name
      description
      acronym
      status
      businessOwnerOrg
      businessOwnerOrgComp
      systemMaintainerOrg
      systemMaintainerOrgComp
    }
  }
`;
