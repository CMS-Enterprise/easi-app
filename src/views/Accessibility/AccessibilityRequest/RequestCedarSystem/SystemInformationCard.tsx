import React from 'react';
import { useTranslation } from 'react-i18next';
import { useQuery } from '@apollo/client';
import {
  Button,
  Card,
  CardBody,
  CardFooter,
  CardGroup,
  CardHeader,
  IconArrowForward,
  IconBookmark
} from '@trussworks/react-uswds';
import classNames from 'classnames';

import UswdsReactLink from 'components/LinkWrapper';
import Alert from 'components/shared/Alert';
import Spinner from 'components/Spinner';
import GetCedarSystemQuery from 'queries/GetCedarSystemQuery';
import {
  GetCedarSystem,
  GetCedarSystemVariables
} from 'queries/types/GetCedarSystem';

type SystemInformationCardProps = {
  cedarSystemId: string;
  bookmarked: boolean;
  toggleCedarSystemBookmark: () => void;
};

const SystemInformationCard = ({
  cedarSystemId,
  bookmarked,
  toggleCedarSystemBookmark
}: SystemInformationCardProps) => {
  const { t } = useTranslation('accessibility');

  const { data, loading, error } = useQuery<
    GetCedarSystem,
    GetCedarSystemVariables
  >(GetCedarSystemQuery, {
    variables: { id: cedarSystemId }
  });

  const toggleBookmark = () => {
    toggleCedarSystemBookmark();
  };

  if (loading)
    return (
      <Spinner
        size="large"
        aria-valuetext={t('newRequestForm.loadingSystems')}
        aria-busy
        data-testid="cedar-system-loading"
      />
    );

  if (error)
    return <Alert type="error">{t('newRequestForm.errorSystemInfo')}</Alert>;

  return (
    <CardGroup>
      <Card>
        <CardHeader className="display-flex">
          <div>
            <h4 className="margin-y-0 line-height-sans-5">
              <UswdsReactLink to={`/systems/${cedarSystemId}`}>
                {data?.cedarSystem?.name}
              </UswdsReactLink>
            </h4>
            <div className="font-sans-xs line-height-sans-1 margin-top-1">
              {data?.cedarSystem?.acronym}
            </div>
          </div>
          <div>
            <Button
              type="button"
              unstyled
              className={classNames('margin-left-105', {
                'text-base-lighter': !bookmarked,
                'text-primary': bookmarked
              })}
              onClick={() => toggleBookmark()}
            >
              <IconBookmark size={3} />
            </Button>
          </div>
        </CardHeader>
        <CardBody>
          <div className="system-info-card__description">
            {data?.cedarSystem?.description}
          </div>
        </CardBody>
        <CardFooter>
          <UswdsReactLink to={`/systems/${cedarSystemId}`}>
            {t('requestDetails.viewSystemProfile')}
            <IconArrowForward className="margin-left-05 margin-bottom-2px text-tbottom" />
          </UswdsReactLink>
        </CardFooter>
      </Card>
    </CardGroup>
  );
};

export default SystemInformationCard;
