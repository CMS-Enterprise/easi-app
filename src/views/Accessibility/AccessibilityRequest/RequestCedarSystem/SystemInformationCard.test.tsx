import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { MockedProvider } from '@apollo/client/testing';
import { render, waitForElementToBeRemoved } from '@testing-library/react';

import GetCedarSystemQuery from 'queries/GetCedarSystemQuery';

import SystemInformationCard from './SystemInformationCard';

const cedarSystemQuery = {
  request: {
    query: GetCedarSystemQuery,
    variables: {
      id: '000-0000-1'
    }
  },
  result: {
    data: {
      cedarSystem: {
        id: '000-0000-1',
        name: 'CMS Operations Information Network',
        description:
          'COINâ€™s goal is to be the secure, role-based single-source provider of business operations data brought together from CMS systems into one easy-to-use interface for stakeholders to drive business decisions. COIN is operated jointly by OSPR and OIT.',
        acronym: 'COIN',
        status: '',
        businessOwnerOrg: '',
        businessOwnerOrgComp: '',
        systemMaintainerOrg: '',
        systemMaintainerOrgComp: ''
      }
    }
  }
};

function renderSystemInformationCard(query: any) {
  return render(
    <MemoryRouter initialEntries={['/508/requests/000-0000-1/cedar-system']}>
      <MockedProvider mocks={[query]}>
        <SystemInformationCard
          cedarSystemId="000-0000-1"
          bookmarked={false}
          toggleCedarSystemBookmark={() => {}}
        />
      </MockedProvider>
    </MemoryRouter>
  );
}

describe('SystemInformationCard', () => {
  it('renders after request', async () => {
    const { asFragment, getByTestId } = renderSystemInformationCard(
      cedarSystemQuery
    );

    await waitForElementToBeRemoved(() => getByTestId('cedar-system-loading'));

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders an error on request fail', async () => {
    const queryError = {
      request: cedarSystemQuery.request,
      error: new Error('bad request')
    };
    const { getByTestId } = renderSystemInformationCard(queryError);

    await waitForElementToBeRemoved(() => getByTestId('cedar-system-loading'));

    expect(getByTestId('alert')).toMatchSnapshot();
  });
});
