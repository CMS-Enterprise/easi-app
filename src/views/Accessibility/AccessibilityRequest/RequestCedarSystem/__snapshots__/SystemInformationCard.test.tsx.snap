// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SystemInformationCard renders after request 1`] = `
<DocumentFragment>
  <ul
    class="usa-card-group"
    data-testid="CardGroup"
  >
    <li
      class="usa-card"
      data-testid="Card"
    >
      <div
        class="usa-card__container"
      >
        <header
          class="usa-card__header display-flex"
          data-testid="CardHeader"
        >
          <div>
            <h4
              class="margin-y-0 line-height-sans-5"
            >
              <a
                class="usa-link"
                href="/systems/000-0000-1"
              >
                CMS Operations Information Network
              </a>
            </h4>
            <div
              class="font-sans-xs line-height-sans-1 margin-top-1"
            >
              COIN
            </div>
          </div>
          <div>
            <button
              class="usa-button usa-button--unstyled margin-left-105 text-base-lighter"
              data-testid="button"
              type="button"
            >
              <svg
                class="usa-icon usa-icon--size-3"
                focusable="false"
                height="1em"
                role="img"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M17 3H7c-1.1 0-1.99.9-1.99 2L5 21l7-3 7 3V5c0-1.1-.9-2-2-2z"
                />
              </svg>
            </button>
          </div>
        </header>
        <div
          class="usa-card__body"
          data-testid="CardBody"
        >
          <div
            class="system-info-card__description"
          >
            COINâ€™s goal is to be the secure, role-based single-source provider of business operations data brought together from CMS systems into one easy-to-use interface for stakeholders to drive business decisions. COIN is operated jointly by OSPR and OIT.
          </div>
        </div>
        <div
          class="usa-card__footer"
          data-testid="CardFooter"
        >
          <a
            class="usa-link"
            href="/systems/000-0000-1"
          >
            View system profile
            <svg
              class="usa-icon margin-left-05 margin-bottom-2px text-tbottom"
              focusable="false"
              height="1em"
              role="img"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
              />
            </svg>
          </a>
        </div>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
